docs/proposals/ns-scoped-catalog.md

```diff
@@ -0,0 +1,256 @@
+
+Adding namespaced resources for service brokers (`ClusterServiceBroker`), services (`ClusterServiceClass`) and plans (`ClusterServicePlans`) allows us to take advantage of the existing namespace concept to perform access control.
+
+### Filtering services and plans from a broker
+
+Adding namespaced resources for brokers, services and plans is necessary but not sufficient to control access to services and plans. A single broker may offer a mix of services that all users should be able to access and services that should only be usable by some users.
+
+In order to prevent a broker that offers a mix of unprivileged and privileged services to the cluster-scoped catalog, there must be a way to filter the services and plans exposed by a broker. This can be accomplished through the use of white/black lists that control which services and plans in a broker's catalog have service-catalog resources created for them. For example:
+
+- A cluster administrator should be able to prevent privileged services from appearing in cluster-scoped catalog
+- A cluster administrator should be able to add certain privileged services to a namespace
+
+## Design
+
+In this proposal we'll focus on adding the namespaced `ServiceBroker`, `ServiceClass`, and `ServicePlan` resources. For details on filtering which services and plans in a broker's catalog have k8s resources created for them, see https://github.com/kubernetes-incubator/service-catalog/pull/1773.
+
+### Namespaced resources
+
+The namespaced resources for brokers, services, and plans should have the same behaviors as their cluster-scoped cousins. To a great degree, we can reuse the same API fields in the namespaced resources, but there are some exceptions:
+
+#### `ServiceBroker` resource
+
+The API for the `ServiceBroker` resource should differ from `ClusterServiceBroker` in exactly one area:
+
+- A user should only be able to specify a secret within the same namespace to hold the auth information
+
+#### `ServiceClass` resource
+
+Differences between `ClusterServiceClass` and `ServiceClass`:
+
+- `ServiceClass.Spec` should have `ServiceBrokerName` instead of `ClusterServiceBrokerName`
+
+#### `ServicePlan` resource
+
+Differences between `ClusterServicePlan` and `ServicePlan`
+
+- `ServicePlan.Spec` should have `ServiceBrokerName` instead of `ClusterServiceBrokerName`
+- `ServicePlan.Spec` should have `ServiceClassRef` instead of `ClusterServiceClassRef`
+
+### Changes to `ServiceInstance`
+
+The `ServiceInstance` resource should be changed to allow users to unambiguously specify a `ServiceClass` and `ServicePlan` instead of the cluster-level resources.
+
+- Add fields to `PlanReference` to represent the external and k8s names of `ServiceClass` and `ServicePlan` (as opposed to the cluster-scoped versions)
+- Add reference fields to `ServiceInstanceSpec` that represent the references to the namespaced resources
```

Comment:
It's taken me this far, but I think we're trying to avoid requiring the user to specify the namespace, as "obviously the namespace is X, that is where we're doing all of these operations".

Reply:
That's right - references within a namespace don't have the namespace specified.